### PR TITLE
Add layer ID to tortoise opinion map (instead of #2734)

### LINKED ISSUE #2424: Add exception votes against blocks the base block votes for
## Description
In the verifying tortoise, when selecting a base block, a node considers the set of blocks in every layer against the "local opinion" for that layer. It adds exceptions to vote _for_ blocks that are in the local opinion but that the base block doesn't explicitly support, and it adds exceptions to vote _against_ blocks that are in the layer but not in the local opinion (which the base block doesn't already explicitly vote against). However, there's currently a flaw in this logic: if the base block supports a block in a given layer that the node hasn't seen--i.e., it's not in the database for that layer and it's not in the local opinion for the layer--then the node should add an exception vote against that block. There's currently no (efficient) way to do this because the base block's opinion vector is just a single long list of the blockIDs that it supports, not indexed by layer. We probably need a better data structure.

## Affected code

This issue appears in commit hash: 007bdb9629ef16687e206950c0a1565476aef070

Related files (optionally with line numbers): `tortoise/verifying_tortoise.go`

https://github.com/spacemeshos/go-spacemesh/blob/43cfaeec21f889d65e2f3f7a5efcadef4834c434/tortoise/verifying_tortoise.go#L480-L488

### LINKED ISSUE #2719: Tortoise vote encoding data structure
Right now tortoise stores/encodes votes in a matrix of size `windowSize**2 * blocksPerLayer**2`: each block includes votes from blocks going back `windowSize` layers, each of which includes votes going back `windowSize` layers.

It seems that this data structure is too big. See @dshulyak comment here: https://github.com/spacemeshos/go-spacemesh/pull/2683#issuecomment-904322215.

Per @tal-m,

> The data structure I had in mind is a 2D matrix, where each row represents a voting block and each column a content block. Each element of the matrix is the opinion of a voting block about the corresponding content block. The window size comes into play because we don't need to store the detailed opinions about content blocks that are more than `windowSize` layers in the past (for those, we only store our own opinion, which should be identical to the opinion of all the voting blocks we see)

> Regarding optimizations:
> I think the unified content block will be a huge improvement to complexity --- since there is almost always 1 block per layer, the matrix needs to be windowSize*votesPerLayer (votesPerLayer=blocksPerLayer in the previous scheme). So that's a blocksPerLayer factor improvement; if we're currently using 50 blocks per layer, it means that even without any additional optimization, the vote storage should be less than 140MB
> I'm sure there's a lot of optimization we can do in the data structure itself. Are we storing it in bit-packed fields? Are we making use of the fact that most honest votes actually have the same opinion? Etc.

See [this slack thread](https://spacemesh.slack.com/archives/CG91F13C7/p1629850014078200)

### PR DESCRIPTION
**Pull request by nkryuchkov**
## Motivation
Based on #2734
Supersedes #2734
Closes #2719 
See also #2679

*Most of the code was copied from #2734 with some fixes. It was hard to fix a couple of tests in #2734 due to some changes that may be excluded from the PR (e.g. https://github.com/spacemeshos/go-spacemesh/pull/2734/files#diff-de1c771222d45ea483c2a6212af7b40473ea6c8c814ecec398deae0ffaddb0d7R1156). Also, uncommenting of https://github.com/spacemeshos/go-spacemesh/pull/2734/files#diff-de1c771222d45ea483c2a6212af7b40473ea6c8c814ecec398deae0ffaddb0d7R491 wasn't copied because it broke `TestLayerPatterns/heal_after_bad_layers`, this likely needs to be done though (#2424).*

## Changes

- Add `layerID` to the inner tortoise block opinion map

This allows efficient eviction of opinions of blocks in older layers,
and it allows us to efficiently add votes against blocks the base block
votes for that are missing from the input vector for a layer

## Test Plan

- Unit tests
- System tests

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

**Comment by nkryuchkov** [comment]
bors try

File: tortoise/state.go
Comment (countvonzero) [inline tortoise/state.go]: i think it's more readable to do
layer1Offset := len(namespaceOpinions)
block1Offset := layer1Offset + types.LayerIDSize
layer2Offset := block1OffSet + types.BlockIDSize
block2Offset := layer2Offset + types.LayerIDSize

File: tortoise/state.go
Comment (countvonzero) [inline tortoise/state.go]: use [block1Offset : layer2Offset]

File: tortoise/state.go
Comment (countvonzero) [inline tortoise/state.go]: use [layer2Offset : block2Offset]

File: tortoise/verifying_tortoise.go
Comment (countvonzero) [inline tortoise/verifying_tortoise.go:484]: are you planning to add against diff in a follow-up PR?

File: tortoise/verifying_tortoise.go
Comment (countvonzero) [inline tortoise/verifying_tortoise.go]: should be able to look up layerID from `state.BlockLayer` first to avoid accessing disk

File: tortoise/verifying_tortoise.go
Comment (countvonzero) [inline tortoise/verifying_tortoise.go]: ditto

File: tortoise/verifying_tortoise.go
Comment (countvonzero) [inline tortoise/verifying_tortoise.go]: ditto

**Review by dshulyak — CHANGES_REQUESTED** [review]
wherever bdp.GetBlock is called in a loop, and there are many outer loops as well, it should be replaced with a call BlockLayer to get a layer from memory

File: tortoise/verifying_tortoise.go
Comment (nkryuchkov) [reply tortoise/verifying_tortoise.go:484]: Either in a follow-up one or in this one if I fix the broken by this addition tests until merging and if the amount of changes is tiny

**Comment by nkryuchkov** [comment]
bors try

File: tortoise/verifying_tortoise.go
Comment (dshulyak) [inline tortoise/verifying_tortoise.go]: please deduplicate this code

**Comment by nkryuchkov** [comment]
bors try-

**Comment by dshulyak** [comment]
i merged this change with https://github.com/spacemeshos/go-spacemesh/pull/2900 and launched a new miner in the cloud to see how it affects rerun, memory and some other things. will update later

**Comment by dshulyak** [comment]
my observation so far is that this doesn't reduce memory by a lot. definitely not 50%, maybe slightly. my node that runs locally uses around 1gb of memory as before. 

rerun became two times slower, and i expected it to be faster. i didn't look into profiles, but this is definitely a problem.

**Comment by dshulyak** [comment]
i suggest postponing this change until the unified content block is introduced, as it very likely will be obsolete. we already considered rewriting the core data structures of a tortoise and it is will happen after a unified content block, if not before. 

i don't like that it makes full rerun ~2x slower and introduces more complexity into the data structure. 
also, i don't see that it makes a substantial difference memory-wise. also it enables https://github.com/spacemeshos/go-spacemesh/issues/2424 , which we will have to support, but we can try inefficient variant first and use until better variant will be found.

**Comment by nkryuchkov** [comment]
@dshulyak I agree. It doesn't enable #2424 yet, I will remove mentioning #2424 and will put this PR on hold then

**Comment by dshulyak** [comment]
tortoise is fully rewritten in https://github.com/spacemeshos/go-spacemesh/pull/2994 , this data structure is not relevant

**Pull request closed without merging**
